src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -4248,26 +4303,17 @@ VOID    MethodTableBuilder::InitializeFieldDescs(FieldDesc *pFieldDescList,
                     }
                 }
 
+                if (fIsStatic && ElementType != ELEMENT_TYPE_VALUETYPE)
+                {
+                    fIsByValue = FALSE; // we're going to treat it as the underlying type now
+                    goto GOT_ELEMENT_TYPE;
+                }
+                
                 // #FieldDescTypeMorph  IF it is an enum, strip it down to its underlying type
-                if (IsSelfRef(pByValueClass) ? IsEnum() : pByValueClass->IsEnum())
-                {
-                    if (IsSelfRef(pByValueClass))
-                    {   // It is self-referencing enum (ValueType) static field - it is forbidden in the ECMA spec, but supported by CLR since v1
-                        // Note: literal static fields are skipped early in this loop
-                        if (bmtMFDescs->ppFieldDescList[0] == NULL)
-                        {   // The field is defined before (the only) instance field
-                            // AppCompat with 3.5 SP1 and 4.0 RTM behavior
-                            BuildMethodTableThrowException(COR_E_BADIMAGEFORMAT, IDS_CLASSLOAD_BAD_FIELD, mdTokenNil);
-                        }
-                        // We will treat the field type as if it was its underlying type (we know its size and will check correctly RVA with the size
-                        // later in this method)
-                        // Therefore we do not have to run code:VerifySelfReferencingStaticValueTypeFields_WithRVA or code:#SelfReferencingStaticValueTypeField_Checks
-                    }
-                    BAD_FORMAT_NOTHROW_ASSERT((IsSelfRef(pByValueClass) ?
-                            bmtEnumFields->dwNumInstanceFields : pByValueClass->GetNumInstanceFields())
-                                == 1); // enums must have exactly one field
-                    FieldDesc * enumField = IsSelfRef(pByValueClass) ?
-                            bmtMFDescs->ppFieldDescList[0] : pByValueClass->GetApproxFieldDescListRaw();
+                if (!fIsStatic && pByValueClass->IsEnum())
+                {
+                    BAD_FORMAT_NOTHROW_ASSERT(pByValueClass->GetNumInstanceFields() == 1); // enums must have exactly one field
```

Comment:
Is it worth encapsulating this logic since we have at least two copies of it? I'm undecided

Reply:
I ended up with no, the logic is short enough that extracting it wasn't very helpful IMO